plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp

```diff
@@ -0,0 +1,226 @@
+      }
+   }
+
+   // Only called at plugin startup
+   template <typename T>
+   void for_each_bp_peer_address(T&& fun) const {
+      for (auto& [_, addr] : config.bp_peer_addresses) { fun(addr); }
+   }
+
+   // Only called from connection strand and the connection constructor
+   void mark_bp_connection(Connection* conn) const {
+      /// mark an connection as a bp connection if it connects to an address in the bp peer list, so that the connection
+      /// won't be subject to the limit of max_client_count.
+
+      if (auto itr = config.bp_peer_accounts.find(conn->log_p2p_address); itr != config.bp_peer_accounts.end()) {
+         conn->is_bp_connection = true;
+      }
+   }
+
+   // Only called from connection strand
+   template <typename Conn>
+   static bool established_client_connection(Conn&& conn) {
+      return !conn->is_bp_connection && conn->socket_is_open() && conn->incoming_and_handshake_received();
+   }
+
+   // Only called from connection strand
+   std::size_t num_established_clients() const {
+      uint32_t num_clients = 0;
+      self()->for_each_connection([&num_clients](auto&& conn) {
+         if (established_client_connection(conn)) {
+            ++num_clients;
+         }
+         return true;
+      });
+      return num_clients;
+   }
+
+   // Only called from connection strand
+   // This should only be called after the first handshake message is received to check if an incoming connection
+   // has exceeded the pre-configured max_client_count limit.
+   bool exceeding_connection_limit(Connection* new_connection) const {
+      return auto_bp_peering_enabled() && self()->max_client_count != 0 &&
+             established_client_connection(new_connection) && num_established_clients() > self()->max_client_count;
+   }
+
+   // Only called from main thread
+   neighbor_finder_type neighbor_finder(const std::vector<chain::producer_authority>& schedule) const {
+      return neighbor_finder_type(config, schedule);
+   }
+
+   // Only called from main thread
+   void on_pending_schedule(const chain::producer_authority_schedule& schedule) {
+      if (auto_bp_peering_enabled() && self()->in_sync()) {
+         if (schedule.producers.size()) {
+            if (pending_schedule_version != schedule.version) {
+               /// establish connection to the BPs within our pending scheduling proximity
+
+               fc_dlog(self()->get_logger(), "pending producer schedule switches from version ${old} to ${new}",
+                       ("old", pending_schedule_version)("new", schedule.version));
+
+               auto finder                       = neighbor_finder(schedule.producers);
+               auto pending_downstream_neighbors = finder.downstream_neighbors();
+
+               fc_dlog(self()->get_logger(), "pending_downstream_neighbors: ${pending_downstream_neighbors}",
+                       ("pending_downstream_neighbors", to_string(pending_downstream_neighbors)));
```

Comment:
Is there a reason to only log downstream neighbors and not upstream neighbors?

Reply:
Because those the only peers to make connections to.